### PR TITLE
Add OpenAPI schema generation pipeline and derive TypeScript types from it

### PR DESCRIPTION
**Pull request by TheFehr**
## Summary by CodeRabbit

* **New Features**
  * Added automated OpenAPI-based TypeScript type generation to keep the frontend synced with the API.
  * CI now verifies OpenAPI docs and regenerated frontend types are up to date.
* **Bug Fixes**
  * Improved JSON serialization for `BigDecimal` values (now emitted as JSON numbers).
  * Fixed timing-track updates to correctly wait for effort updates.
* **API Changes**
  * Removed `deleted_at` from many JSON responses.
  * Updated invoice breakdown/offer & project payload shapes (including boolean `rate_unit_is_time`) and added invoice fields: `beginning`, `ending`, `project_id`.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -64,8 +64,30 @@ jobs:
       - name: Run Tests
         run: docker compose -f docker-compose.semaphore.yml run -T apir bin/test
 
+      - name: Generate OpenAPI schema
+        run: docker compose -f docker-compose.semaphore.yml run -T -e OPENAPI=1 apir bundle exec rspec spec/requests/
```
[the remark this answers is not in the document]
Comment (TheFehr) [reply .github/workflows/ci.yml]: Already fixed — `--order defined` was added in commit 42d1f897.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -64,8 +64,30 @@ jobs:
...
+
+      - uses: actions/setup-node@v4
+        with:
+          node-version: '20'
+
+      - name: Generate TypeScript types from OpenAPI schema
+        run: cd frontend && npm run generate:types
+
+      - name: Verify generated types are committed
+        run: git diff --exit-code frontend/src/types/generated.ts
```
[the remark this answers is not in the document]
Comment (TheFehr) [reply .github/workflows/ci.yml:90]: Out of scope for this PR. The `actions/checkout@v2` usage is pre-existing across all CI jobs, not something introduced here. Worth addressing in a separate hardening PR.

File: apir/app/views/v2/customers/show.json.jbuilder [hunk 3]
```diff
@@ -1,23 +1,22 @@
 # frozen_string_literal: true
 
-json.extract! @customer, :id, :type, :comment, :company_id, :department, :department_in_address, :email, :first_name, :last_name, :hidden, :archived, :name, :accountant_id, :rate_group_id, :salutation, :created_at, :updated_at, :deleted_at
+json.extract! @customer, :id, :type, :comment, :company_id, :department, :department_in_address, :email, :first_name, :last_name, :hidden, :archived, :name, :accountant_id, :rate_group_id, :salutation, :created_at, :updated_at
 
 json.company do
   if @customer.company
-    json.extract! @customer, :id, :type, :comment, :company_id, :department, :email, :department_in_address, :first_name, :last_name, :hidden, :archived, :name, :accountant_id, :rate_group_id, :salutation, :created_at, :updated_at,
-                  :deleted_at
+    json.extract! @customer, :id, :type, :comment, :company_id, :department, :email, :department_in_address, :first_name, :last_name, :hidden, :archived, :name, :accountant_id, :rate_group_id, :salutation, :created_at, :updated_at
```
[the remark this answers is not in the document]
Comment (TheFehr) [reply apir/app/views/v2/customers/show.json.jbuilder]: Fixed in 4920a9c1 — now correctly extracts from `@customer.company`.

File: apir/app/views/v2/offers/_offer.json.jbuilder [hunk 4]
```diff
@@ -9,22 +9,31 @@ json.project_id offer.project.id if offer.project
...
+end
 json.positions offer.offer_positions.sort_by(&:order) do |position|
   json.extract! position, :id, :amount, :description, :price_per_rate, :rate_unit_id, :service_id,
-                :vat, :order, :position_group_id, :service, :rate_unit_archived
+                :vat, :order, :position_group_id, :rate_unit_archived
+  json.service do
+    if position.service
+      json.extract! position.service, :id, :name, :description, :vat, :order, :local_order, :archived, :service_category_id
+    end
+  end
```
[the remark this answers is not in the document]
Comment (TheFehr) [reply apir/app/views/v2/offers/_offer.json.jbuilder:34]: Low priority — the `{}` empty object doesn't cause runtime crashes since JS property access on `{}` returns `undefined`. The frontend already uses optional chaining in the spots that hit nil services (`p.service?.order || 0`). Changing this would also require updating the OpenAPI test data to cover the nil case for the schema to reflect nullability correctly.

File: apir/app/views/v2/people/show.json.jbuilder [hunk 5]
```diff
@@ -1,22 +1,21 @@
...
 
-json.extract! @person, :id, :type, :comment, :company_id, :department, :department_in_address, :email, :first_name, :last_name, :hidden, :archived, :name, :accountant_id, :rate_group_id, :salutation, :created_at, :updated_at, :deleted_at
+json.extract! @person, :id, :type, :comment, :company_id, :department, :department_in_address, :email, :first_name, :last_name, :hidden, :archived, :name, :accountant_id, :rate_group_id, :salutation, :created_at, :updated_at
 json.company do
   if @person.company
-    json.extract! @person, :id, :type, :comment, :company_id, :department, :department_in_address, :email, :first_name, :last_name, :hidden, :archived, :name, :accountant_id, :rate_group_id, :salutation, :created_at, :updated_at,
-                  :deleted_at
+    json.extract! @person, :id, :type, :comment, :company_id, :department, :department_in_address, :email, :first_name, :last_name, :hidden, :archived, :name, :accountant_id, :rate_group_id, :salutation, :created_at, :updated_at
   end
 end
```
[the remark this answers is not in the document]
Comment (TheFehr) [reply apir/app/views/v2/people/show.json.jbuilder:8]: Fixed in 4920a9c1 — now correctly extracts from `@person.company`.

File: apir/app/views/v2/projects/_project.json.jbuilder [hunk 6]
```diff
@@ -32,7 +34,12 @@ end
 json.positions project.project_positions.sort_by(&:order) do |position|
   json.extract! position, :id, :description, :price_per_rate, :rate_unit_id, :service_id,
                 :vat, :order, :position_group_id, :efforts_value_with_unit, :charge,
-                :rate_unit_archived, :deletable, :service, :is_time
+                :rate_unit_archived, :deletable, :is_time
+  json.service do
+    if position.service
+      json.extract! position.service, :id, :name, :description, :vat, :order, :local_order, :archived, :service_category_id
+    end
+  end
```
[the remark this answers is not in the document]
Comment (TheFehr) [reply apir/app/views/v2/projects/_project.json.jbuilder:42]: Same reasoning as the offers service comment — the frontend already guards with optional chaining, and the empty `{}` doesn't cause crashes. Low priority.

File: apir/config/initializers/bigdecimal_serializer.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+# BigDecimal serializes to JSON as a string by default to avoid precision loss.
+# Our decimal columns (vat, value, factor, fixed_price, etc.) are all within
+# Float's 15-digit precision, so serializing as Float is safe and produces
+# proper JSON numbers instead of strings.
+BigDecimal.class_eval do
+  def as_json(_options = nil)
+    to_f
+  end
+end
```
[the remark this answers is not in the document]
Comment (TheFehr) [reply apir/config/initializers/bigdecimal_serializer.rb:11]: Intentional — the inline comment explains: "Our decimal columns (vat, value, factor, fixed_price, etc.) are all within Float's 15-digit precision, so serializing as Float is safe and produces proper JSON numbers instead of strings." Switching to string serialization would break the frontend, which expects numeric values.

File: apir/spec/rails_helper.rb [hunk 8]
```diff
@@ -38,10 +38,29 @@
...
+  config.before(:suite) do
+    connection = ActiveRecord::Base.connection
+    connection.execute('SET FOREIGN_KEY_CHECKS = 0')
+    connection.tables.each do |table|
+      connection.execute("TRUNCATE TABLE `#{table}`")
+    rescue StandardError
+      nil
+    end
+    connection.execute('SET FOREIGN_KEY_CHECKS = 1')
+  end
```
[the remark this answers is not in the document]
Comment (TheFehr) [reply apir/spec/rails_helper.rb]: Invalid — there is no `before(:suite)` block in `rails_helper.rb`. The code referenced (FK check toggling, table truncation with rescue) does not exist in this file.

File: apir/spec/requests/position_groups_spec.rb [hunk 9]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+# These specs exist solely to generate the OpenAPI schema via rspec-openapi.
+# They are not regression tests — do not add assertions here.
+require 'rails_helper'
+
+# PositionGroups only exposes a create endpoint; there are no index or show routes.
+RSpec.describe 'V2::PositionGroups', type: :request do
+end
```
[the remark this answers is not in the document]
Comment (TheFehr) [reply apir/spec/requests/position_groups_spec.rb:19]: Already done — the spec already contains a `POST /v2/position_groups` example (lines 13–18).

File: frontend/package.json [hunk 10]
```diff
@@ -47,7 +47,8 @@
     "test": "react-scripts-ts test --env=jsdom",
     "eject": "react-scripts-ts eject",
     "format": "tslint --project . -t verbose --fix",
-    "lint": "tslint --project . -t verbose --config ./tslint.prod.json"
+    "lint": "tslint --project . -t verbose --config ./tslint.prod.json",
+    "generate:types": "npx --yes openapi-typescript@7 ../apir/doc/openapi.yaml -o src/types/generated.ts"
```
[the remark this answers is not in the document]
Comment (TheFehr) [reply frontend/package.json:51]: Partially addressed in 4920a9c1 — `generate-types.sh` now calls `npm run generate:types` instead of bare `npx`, so both invocation paths use the same command. The `@7` major version pin in `package.json` is sufficient for a dev-time code generator; an exact pin would just create update churn.

File: .github/workflows/ci.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (TheFehr) [reply .github/workflows/ci.yml:90]: Yes let's open an issue for this

File: apir/spec/rails_helper.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (TheFehr) [reply apir/spec/rails_helper.rb]: Correction to my earlier reply: the `before(:suite)` block did exist in an earlier commit on this branch (f07fbbe) but was removed in a subsequent commit (97a8919). The feedback was valid at the time of review — the code just no longer exists on the branch, so there's nothing to fix.

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -14,7 +14,7 @@ jobs:
       - uses: actions/checkout@v2
 
       - name: Frontend Build
-        run: docker run --rm -v ./frontend:/var/frontend -w /var/frontend --entrypoint "" node:12 bash -c "yarn install && yarn run lint"
+        run: docker run --rm -v ./frontend:/var/frontend -w /var/frontend --entrypoint "" node:20 bash -c "yarn install && yarn run lint"
```
[the remark this answers is not in the document]
Comment (TheFehr) [reply .github/workflows/ci.yml:17]: Hardening — out of scope for this PR. @coderabbitai please add this to #678.

File: frontend/package.json [hunk 12]
```diff
@@ -42,18 +42,19 @@
...
   },
   "devDependencies": {
     "@sentry/typescript": "^4.0.1",
     "@types/history": "^4.7.0",
     "@types/jest": "^23.3.4",
-    "@types/node": "^10.11.5",
+    "@types/node": "^20.0.0",
```
[the remark this answers is not in the document]
Comment (TheFehr) [reply frontend/package.json:57]: Not an issue — the lockfile resolves `typescript@^3.1.1` to 3.9.10, which is compatible with `@types/node@^20`. The `^3.1.1` in package.json is a minimum, not the installed version.

File: frontend/src/form/dialog/PositionMoveDialog.tsx [hunk 13]
```diff
@@ -50,6 +50,8 @@ export async function resolveNewGroupName(newGroupName: string | null, groupingE
       groupingEntity.position_groupings.push({
         id: positionGroupStore!.positionGroup!.id,
         name: positionGroupName,
+        order: groupingEntity.position_groupings.length,
```
[the remark this answers is not in the document]
Comment (TheFehr) [reply frontend/src/form/dialog/PositionMoveDialog.tsx]: Good catch — fixed, now uses `groupingEntity.position_groupings.length + 1` for 1-based ordering.

File: frontend/src/views/employees/employeeSchema.tsx [hunk 14]
```diff
@@ -69,24 +70,24 @@ export const newEmployeeSchema = localizeSchema(() =>
   }),
 );
 
-export const employeeTemplate = {
+// Form initial state — employee_group_id/group/group_name are null until the user selects a group.
+export const employeeTemplate: Omit<Employee, 'employee_group_id' | 'group' | 'group_name'> & { employee_group_id: number | null; group: Employee['group'] | null; group_name: string | null; password?: string } = {
```
[the remark this answers is not in the document]
Comment (TheFehr) [reply frontend/src/views/employees/employeeSchema.tsx:74]: Not an issue — `Omit` was added in TypeScript 3.5, and the lockfile resolves `typescript@^3.1.1` to 3.9.10. The project compiles fine.

**Comment by TheFehr** [comment]
@coderabbitai Regarding the outside-diff comments from the latest review:

**CI permissions (ci.yml lines 10-26)** and **Dockerfile non-root user (prod.Dockerfile)**: Both are valid hardening items but out of scope for this PR. Please add these to #678.

**Double cast `as unknown as Employee` (EmployeeCreate.tsx)** and **`as unknown as Person` (PersonCreate.tsx)**: Intentional. The templates have deliberately relaxed types (nullable fields that are required on the full type) for form initial state. The double cast is the correct pattern here — "fixing" it would mean refactoring the form component prop types, which is unrelated scope.

**Comment by TheFehr** [comment]
@coderabbitai review
